File: packages/python/goldenpipe/goldenpipe/repair_host.py
"""Host glue for repair-plan: sample polars columns, build ColumnInputs, call the
pure kernel, attach the advisory artifact + reasoning. Advisory ONLY — never
mutates the stage list."""
from __future__ import annotations

from goldenpipe.repair import build_repair_plan

_SAMPLE_LIMIT = 20


def sample_column(series, limit: int = _SAMPLE_LIMIT) -> list[str]:
    out: list[str] = []
    for v in series:                       # polars Series iterates values in row order
        if v is None:
            continue
        s = str(v)
        if s.strip() == "":
            continue
        out.append(s)
        if len(out) >= limit:
            break
    return out


def _coarse_str(t) -> str:
    # ColumnType is `str, Enum` with lowercase values, BUT str(ColumnType.EMAIL) ==
    # "ColumnType.EMAIL" (NOT "email") — only .value gives "email". The context
    # builder can also leave inferred_type a raw str (column_context.py coercion),
    # which has no .value. getattr handles both.
    return getattr(t, "value", t)


def build_column_inputs(contexts, df) -> list[dict]:
    cols = []
    names = set(df.columns)
    for ctx in contexts:
        if ctx.name not in names:
            continue
        cols.append({
            "name": ctx.name,
            "coarse_type": _coarse_str(ctx.inferred_type),
            "samples": sample_column(df[ctx.name]),
        })
    return cols


def attach_repair_plan(ctx, findings, contexts, df) -> dict:
    columns = build_column_inputs(contexts, df)
    plan = build_repair_plan(findings, columns)
    ctx.artifacts["repair_plan"] = plan
    lines = [
        f"repair: {item['column']} ({item['check']}) -> "
        f"{','.join(item['suggested_transforms'])} [{item['reason']}]"
        for item in plan["repairs"]
    ]
    if lines:
        ctx.reasoning["repair_plan"] = "\n".join(lines)
    return plan


# ── Phase 2: active application (fixer allowlist + config conversion) ─────
# FIXERS are transforms that CLEAN in place. Everything else the kernel can
# suggest is a *_validate ASSERTION that returns a boolean series — applying it
# as a column op would overwrite the column with True/False, so it is skipped.
# Host policy, kept identical in repairHost.ts (NOT the parity-gated kernel).
FIXERS = frozenset({
    "fix_mojibake", "normalize_unicode", "date_parse", "email_normalize",
    "email_canonical", "name_proper", "phone_national", "zip_normalize",
})


def repair_transform_specs(plan: dict) -> tuple[list[dict], list[dict]]:
    """repair_plan -> (specs, skipped). specs = [{column, ops}] fixer ops grouped
    per column, deduped, order-preserving. skipped = [{column, op}] assertion ops."""
    by_col: dict[str, list[str]] = {}
    order: list[str] = []
    skipped: list[dict] = []
    for item in plan.get("repairs", []):
        col = item["column"]
        for op in item.get("suggested_transforms", []):
            if op in FIXERS:
                if col not in by_col:
                    by_col[col] = []
                    order.append(col)
                if op not in by_col[col]:
                    by_col[col].append(op)
            else:
                skipped.append({"column": col, "op": op})
    specs = [{"column": c, "ops": by_col[c]} for c in order]
    return specs, skipped


def merge_transforms(user: list[dict], repair: list[dict]) -> list[dict]:
    """Per-column merge, user ops first then repair ops, dedup exact dupes,
    preserving first-seen column + op order."""
    by_col: dict[str, list[str]] = {}
    order: list[str] = []
    for spec in list(user) + list(repair):
        col = spec["column"]
        if col not in by_col:
            by_col[col] = []
            order.append(col)
        by_col[col].extend(spec.get("ops") or [])
    result = []
    for col in order:
        seen: set[str] = set()
        ops: list[str] = []
        for op in by_col[col]:
            if op not in seen:
                seen.add(op)
                ops.append(op)
        result.append({"column": col, "ops": ops})
    return result

File: packages/typescript/goldenpipe/src/core/repairHost.ts
/**
 * repairHost.ts — TS producer glue for the repair plan (mirror of
 * goldenpipe/repair_host.py). Samples column values from the row array, builds
 * ColumnInputs, calls the pure repair.ts kernel, attaches the advisory artifact.
 * Also holds the Phase-2 FIXERS allowlist + config conversion (host policy,
 * identical to repair_host.py — NOT the parity-gated kernel).
 */
import { buildRepairPlan } from "./repair.js";
import type { ColumnInput, Finding, RepairPlan } from "./repair.js";
import type { ColumnContext } from "./columnContext.js";
import type { PipeContext, Row } from "./models.js";
import type { TransformSpec } from "goldenflow/core";

const SAMPLE_LIMIT = 20;

export function sampleColumn(rows: Row[], col: string, limit = SAMPLE_LIMIT): string[] {
  const out: string[] = [];
  for (const row of rows) {
    const v = (row as Record<string, unknown>)[col];
    if (v === null || v === undefined) continue;
    const s = String(v);
    if (s.trim() === "") continue;
    out.push(s);
    if (out.length >= limit) break;
  }
  return out;
}

export function buildColumnInputs(contexts: ColumnContext[], rows: Row[]): ColumnInput[] {
  const first = rows.length > 0 ? rows[0] : undefined;
  const names = new Set<string>(first ? Object.keys(first as object) : []);
  const cols: ColumnInput[] = [];
  for (const ctx of contexts) {
    if (!names.has(ctx.name)) continue;
    cols.push({ name: ctx.name, coarse_type: ctx.inferredType, samples: sampleColumn(rows, ctx.name) });
  }
  return cols;
}

export function attachRepairPlan(
  ctx: PipeContext,
  findings: Finding[],
  contexts: ColumnContext[],
  rows: Row[],
): RepairPlan {
  const columns = buildColumnInputs(contexts, rows);
  const plan = buildRepairPlan(findings, columns);
  ctx.artifacts["repair_plan"] = plan;
  const lines = plan.repairs.map(
    (item) => `repair: ${item.column} (${item.check}) -> ${item.suggested_transforms.join(",")} [${item.reason}]`,
  );
  if (lines.length > 0) ctx.reasoning["repair_plan"] = lines.join("\n");
  return plan;
}

// ── FIXERS allowlist + conversion (identical to repair_host.py) ──────────
export const FIXERS: ReadonlySet<string> = new Set([
  "fix_mojibake", "normalize_unicode", "date_parse", "email_normalize",
  "email_canonical", "name_proper", "phone_national", "zip_normalize",
]);

export function repairTransformSpecs(plan: RepairPlan | undefined): {
  specs: TransformSpec[];
  skipped: Array<{ column: string; op: string }>;
} {
  const byCol = new Map<string, string[]>();
  const order: string[] = [];
  const skipped: Array<{ column: string; op: string }> = [];
  for (const item of plan?.repairs ?? []) {
    const col = item.column;
    for (const op of item.suggested_transforms ?? []) {
      if (FIXERS.has(op)) {
        let ops = byCol.get(col);
        if (ops === undefined) { ops = []; byCol.set(col, ops); order.push(col); }
        if (!ops.includes(op)) ops.push(op);
      } else {
        skipped.push({ column: col, op });
      }
    }
  }
  const specs: TransformSpec[] = order.map((c) => ({ column: c, ops: byCol.get(c) ?? [] }));
  return { specs, skipped };
}

export function mergeTransforms(user: readonly TransformSpec[], repair: readonly TransformSpec[]): TransformSpec[] {
  const byCol = new Map<string, string[]>();
  const order: string[] = [];
  for (const spec of [...user, ...repair]) {
    let ops = byCol.get(spec.column);
    if (ops === undefined) { ops = []; byCol.set(spec.column, ops); order.push(spec.column); }
    ops.push(...spec.ops);
  }
  return order.map((col) => {
    const seen = new Set<string>();
    const ops: string[] = [];
    for (const op of byCol.get(col) ?? []) if (!seen.has(op)) { seen.add(op); ops.push(op); }
    return { column: col, ops };
  });
}
